### PR TITLE
[GLUTEN-7028][CH][Part-11] Support write parquet files with bucket

### PR DESCRIPTION
**Pull request by lwz9103**
## What changes were proposed in this pull request?

Support write parquet files with bucket

(Fixes: \#7028)

## How was this patch tested?

unit tests

### CODE REVIEW

File: cpp-ch/local-engine/Parser/RelParsers/WriteRelParser.cpp [hunk 1]
```diff
@@ -52,14 +52,15 @@ DB::ProcessorPtr make_sink(
     const DB::Block & input_header,
     const DB::Block & output_header,
     const std::string & base_path,
-    const FileNameGenerator & generator,
+    FileNameGenerator & generator,
```
Comment (baibaichen) [inline cpp-ch/local-engine/Parser/RelParsers/WriteRelParser.cpp:55]: const

File: cpp-ch/local-engine/Storages/Output/NormalFileWriter.h [hunk 2]
```diff
@@ -230,20 +230,57 @@ class WriteStats : public WriteStatsBase
 
 struct FileNameGenerator
 {
-    const bool pattern;
-    const std::string filename_or_pattern;
+    // Align with org.apache.spark.sql.execution.FileNamePlaceHolder
+    const std::vector<std::string> placeholders = {"{id}", "{bucket}"};
```
Comment (baibaichen) [inline cpp-ch/local-engine/Storages/Output/NormalFileWriter.h:234]: 静态变量

File: cpp-ch/local-engine/Storages/Output/NormalFileWriter.h [hunk 3]
```diff
@@ -230,20 +230,57 @@ class WriteStats : public WriteStatsBase
...
+        std::string result = file_pattern;
+        if (need_replaced[0])
+            result = pattern_format(placeholders[0], toString(DB::UUIDHelpers::generateV4()));
+        if (need_replaced[1])
+            result = pattern_format(placeholders[1], bucket);
+        return result;
+    }
 
-    std::string generate() const
+    std::string pattern_format(std::string arg, std::string replacement) const
```
Comment (baibaichen) [inline cpp-ch/local-engine/Storages/Output/NormalFileWriter.h:267]: 参数使用引用

File: cpp-ch/local-engine/Parser/RelParsers/WriteRelParser.cpp (hunk 1 above)
Comment (lwz9103) [reply cpp-ch/local-engine/Parser/RelParsers/WriteRelParser.cpp]: resolved

File: cpp-ch/local-engine/Storages/Output/NormalFileWriter.h (hunk 2 above)
Comment (lwz9103) [reply cpp-ch/local-engine/Storages/Output/NormalFileWriter.h]: resolved

File: cpp-ch/local-engine/Storages/Output/NormalFileWriter.h (hunk 3 above)
Comment (lwz9103) [reply cpp-ch/local-engine/Storages/Output/NormalFileWriter.h]: thanks, resolved

**Review by baibaichen — APPROVED** [review]

**Pull request merged**